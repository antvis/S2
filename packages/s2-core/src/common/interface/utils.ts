export type PickEssential<O> = {
  [K in keyof O as Pick<Partial<O>, K> extends Pick<O, K> ? never : K]: O[K];
};
