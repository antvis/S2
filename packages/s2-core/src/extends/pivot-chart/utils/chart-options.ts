import type { AxisComponent, G2Spec } from '@antv/g2';
import {
  EXTRA_FIELD,
  G2_THEME_TYPE,
  type InternalFullyCellTheme,
  type Node,
  type SpreadSheet,
  type ViewMeta,
} from '@antv/s2';
import { map, unary } from 'lodash';
import { X_FIELD_FORMATTER } from '../constant';
import type { PivotChartSheet } from '../pivot-chart-sheet';

export function getTheme(s2: SpreadSheet): Pick<G2Spec, 'theme'> {
  const themeName = s2.getThemeName();

  return {
    theme: {
      type: G2_THEME_TYPE[themeName] ?? 'light',
    },
  };
}

export function getAxisStyle(cellStyle: InternalFullyCellTheme): AxisComponent {
  return {
    // title
    titleSpacing: 0,
    titleFontSize: cellStyle.bolderText.fontSize,
    titleFontFamily: cellStyle.bolderText.fontFamily,
    titleFontWeight: cellStyle.bolderText.fontWeight,
    titleFill: cellStyle.bolderText.fill,
    titleFillOpacity: cellStyle.bolderText.opacity,

    // label
    labelAlign: 'horizontal',
    labelAutoRotate: false,
    labelFontSize: cellStyle.text.fontSize,
    labelFontFamily: cellStyle.text.fontFamily,
    labelFontWeight: cellStyle.text.fontWeight,
    labelFill: cellStyle.text.fill,
    labelFillOpacity: cellStyle.text.opacity,
    labelStroke: cellStyle.text.fill,
    labelStrokeOpacity: cellStyle.text.fill,

    // tick
    tick: true,
    tickStroke: cellStyle.text.fill,
    tickStrokeOpacity: cellStyle.text.opacity,

    // line
    line: false,
    lineStroke: cellStyle.text.fill,
    lineStrokeOpacity: cellStyle.text.opacity,

    // grid
    grid: false,
  };
}

export function getCoordinate(s2: SpreadSheet): Pick<G2Spec, 'coordinate'> {
  if ((s2 as PivotChartSheet).isPolarCoordinate?.()) {
    return {};
  }

  return {
    coordinate: {
      transform: s2.isValueInCols() ? [{ type: 'transpose' }] : undefined,
    },
  };
}

export function getAxisXOptions(meta: Node, s2: SpreadSheet): AxisComponent {
  const domain = map(meta.children, (child) => {
    const formatter = s2.dataSet.getFieldFormatter(child.field);

    return !child.isTotalRoot && formatter
      ? formatter(child.value, undefined, child)
      : child.value;
  });

  return {
    type: 'axisX',
    scale: {
      x: {
        type: 'band',
        domain,
        range: [0, 1],
      },
    },
  };
}

export function getScaleY(
  value: string,
  s2: SpreadSheet,
): Pick<G2Spec, 'scale'> {
  if ((s2 as PivotChartSheet).isPolarCoordinate?.()) {
    return {};
  }

  const range = s2.dataSet.getValueRangeByField(value);

  return {
    scale: {
      y: {
        type: 'linear',
        domain: [range.minValue, range.maxValue],
        range: [1, 0],
      },
    },
  };
}

export function getAxisYOptions(meta: Node, s2: SpreadSheet): AxisComponent {
  const { field, value } = meta;

  const formatter = s2.dataSet.getFieldFormatter(value);

  return {
    type: 'axisY',
    ...getScaleY(value, s2),
    labelFormatter: unary(formatter),
    title: s2.dataSet.getFieldFormatter(field)?.(value),
  };
}

export function getTooltip(
  viewMeta: ViewMeta,
  s2: SpreadSheet,
): Pick<G2Spec, 'tooltip'> {
  const { xField, yField } = viewMeta;
  const dataSet = s2.dataSet;

  return {
    tooltip: {
      title: (data: any) => {
        return data[X_FIELD_FORMATTER]
          ? dataSet.getFieldFormatter(xField!)?.(data[xField!])
          : data[xField!];
      },
      items: [
        (data: any) => {
          return {
            name: dataSet.getFieldFormatter(EXTRA_FIELD)?.(yField),
            value: dataSet.getFieldFormatter(yField!)?.(data[yField!]),
          };
        },
      ],
    },
  };
}
