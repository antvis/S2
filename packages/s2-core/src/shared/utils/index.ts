export * from './classnames';
export * from './drill-down';
export * from './options';
export * from './resize';
