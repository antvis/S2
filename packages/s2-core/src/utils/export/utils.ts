import { concat, get } from 'lodash';
import {
  CopyMIMEType,
  type CopyAllDataParams,
  type Copyable,
  type CopyableItem,
} from '../../common/interface/export';
import { asyncProcessAllSelected } from './copy/core';

/**
 * 同步复制
 */
export const copyToClipboardByExecCommand = (data: Copyable): Promise<void> =>
  new Promise((resolve, reject) => {
    let content: string;

    if (Array.isArray(data)) {
      content = get(
        data.filter((item) => item.type === CopyMIMEType.PLAIN),
        '[0].content',
        '',
      ) as string;
    } else {
      content = data.content || '';
    }

    const textarea = document.createElement('textarea');

    textarea.value = content;
    document.body.appendChild(textarea);
    // 开启 preventScroll, 防止页面有滚动条时触发滚动
    textarea.focus({ preventScroll: true });
    textarea.select();

    const success = document.execCommand('copy');

    document.body.removeChild(textarea);

    if (success) {
      resolve();
    } else {
      reject();
    }
  });

/**
 * 异步复制
 */
export const copyToClipboardByClipboard = (data: Copyable): Promise<void> =>
  navigator.clipboard
    .write([
      new ClipboardItem(
        concat(data).reduce((prev, copyable: CopyableItem) => {
          const { type, content } = copyable;
          // eslint-disable-next-line no-control-regex
          const contentToCopy = content.replace(/\x00/g, '');

          return {
            ...prev,
            [type]: new Blob([contentToCopy], { type }),
          };
        }, {}),
      ),
    ])
    .catch(() => copyToClipboardByExecCommand(data));

/**
 * @name 复制数据
 * @param data 数据源
 * @param sync 同步复制
 */
export const copyToClipboard = (
  data: Copyable | string,
  sync = false,
): Promise<void> => {
  let copyableItem: Copyable;

  if (typeof data === 'string') {
    copyableItem = {
      content: data,
      type: CopyMIMEType.PLAIN,
    };
  } else {
    copyableItem = data;
  }

  if (!navigator.clipboard || !window.ClipboardItem || sync) {
    return copyToClipboardByExecCommand(copyableItem);
  }

  return copyToClipboardByClipboard(copyableItem);
};

/**
 * @name 导出数据
 * @param dataString 数据源 (字符串)
 * @param fileName 文件名 (格式为 csv)
 */
export const download = (dataString: string, fileName: string) => {
  try {
    const link = document.createElement('a');

    link.download = `${fileName}.csv`;
    // Avoid errors in Chinese encoding.
    const dataBlob = new Blob([`\ufeff${dataString}`], {
      type: 'text/csv;charset=utf-8',
    });

    link.href = URL.createObjectURL(dataBlob);
    link.click();
    URL.revokeObjectURL(link.href);
  } catch (error) {
    // eslint-disable-next-line no-console
    console.error(error);
  }
};

/**
 * 异步获取文本数据
 * @example
    const data = await asyncGetAllPlainData({
      sheetInstance: s2,
      split: '\t',
      formatOptions: true,
    });
 */
export const asyncGetAllPlainData = async (params: CopyAllDataParams) => {
  const result = await asyncProcessAllSelected(params);

  return result[0].content;
};
