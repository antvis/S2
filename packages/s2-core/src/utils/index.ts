export { splitTotal } from './data-set-operate';
export { auto } from './formatter';
export * from './layout';
export * from './text';
export * from './theme';
export * from './export/index';
export * from './export/copy';
export * from './interaction';
export * from './g-renders';
export * from './merge';
export * from './is-mobile';
export * from './tooltip';
export * from './get-classnames';
export * from './cell/cell';
export * from './cell/data-cell';
export * from './sort-action';
