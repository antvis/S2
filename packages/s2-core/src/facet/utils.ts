import { SimpleBBox, IGroup } from '@antv/g-canvas';
import { findIndex, isNil } from 'lodash';

import { Indexes } from '../utils/indexes';
import { ViewCellHeights } from './layout/interface';
import {
  FrozenCellType,
  FrozenOpts,
  FrozenCellIndex,
} from '@/common/constant/frozen';
import { Pagination, ScrollSpeedRatio } from '@/common/interface';

export const isFrozenCol = (colIndex: number, frozenCount: number) => {
  return frozenCount > 0 && colIndex < frozenCount;
};

export const isFrozenTrailingCol = (
  colIndex: number,
  frozenCount: number,
  colLength: number,
) => {
  return frozenCount > 0 && colIndex >= colLength - frozenCount;
};

export const isFrozenRow = (
  rowIndex: number,
  minRowIndex: number,
  frozenCount: number,
) => {
  return frozenCount > 0 && rowIndex < minRowIndex + frozenCount;
};

export const isFrozenTrailingRow = (
  rowIndex: number,
  maxRowIndex: number,
  frozenCount: number,
) => {
  return frozenCount > 0 && rowIndex >= maxRowIndex + 1 - frozenCount;
};

/**
 * 计算偏移 scrollX、scrollY 的时候，在视窗中的节点索引
 * @param scrollX
 * @param scrollY
 * @param widths
 * @param heights
 * @param viewport
 * @param rowRemainWidth
 */
export const calculateInViewIndexes = (
  scrollX: number,
  scrollY: number,
  widths: number[],
  heights: ViewCellHeights,
  viewport: SimpleBBox,
  rowRemainWidth?: number,
): Indexes => {
  // 1. 计算 x min、max
  let xMin = findIndex(
    widths,
    (width: number, idx: number) => {
      const x =
        scrollX - (isNil(rowRemainWidth) ? 0 : rowRemainWidth) + viewport.x;
      return x >= width && x < widths[idx + 1];
    },
    0,
  );
  xMin = Math.max(xMin, 0);

  let xMax = findIndex(
    widths,
    (width: number, idx: number) => {
      const x = viewport.width + scrollX + viewport.x;
      return x >= width && x < widths[idx + 1];
    },
    xMin,
  );
  xMax = Math.min(xMax === -1 ? Infinity : xMax, widths.length - 2);

  const { start: yMin, end: yMax } = heights.getIndexRange(
    scrollY + viewport.y,
    viewport.height + scrollY + viewport.y,
  );

  // use direction
  // const halfWidthSize = Math.ceil(xMax - xMin / 4);
  // const halfHeightSize = Math.ceil(yMax - yMin / 4);
  // xMin = Math.max(0, xMin - halfWidthSize)
  // xMax = xMax + halfWidthSize;
  // yMin = Math.max(0, yMin - halfHeightSize);
  // yMax = yMax + halfHeightSize;

  return [xMin, xMax, yMin, yMax];
};

/**
 * 优化滚动方向，对于小角度的滚动，固定为一个方向
 * @param x
 * @param y
 * @param ratio
 */
export const optimizeScrollXY = (
  x: number,
  y: number,
  ratio: ScrollSpeedRatio,
): [number, number] => {
  const ANGLE = 2; // 调参工程师
  const angle = Math.abs(x / y);

  // 经过滚动优化之后的 x, y
  const deltaX = angle <= 1 / ANGLE ? 0 : x;
  const deltaY = angle > ANGLE ? 0 : y;

  return [deltaX * ratio.horizontal, deltaY * ratio.vertical];
};

export const translateGroup = (
  group: IGroup,
  scrollX: number,
  scrollY: number,
) => {
  const matrix = group?.getMatrix();
  const preX = matrix?.[6] ?? 0;
  const preY = matrix?.[7] ?? 0;
  group?.translate(scrollX - preX, scrollY - preY);
};

export const translateGroupX = (group: IGroup, scrollX: number) => {
  const matrix = group?.getMatrix();
  const preX = matrix?.[6] ?? 0;
  group?.translate(scrollX - preX, 0);
};

export const translateGroupY = (group: IGroup, scrollY: number) => {
  const matrix = group?.getMatrix();
  const preY = matrix?.[7] ?? 0;
  group?.translate(0, scrollY - preY);
};

/**
 * frozen                     frozenTrailing
 * ColCount                   ColCount
 * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * |     |     frozenRow     |          |  frozenRowCount
 * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * | fro |                   | fro      |
 * | zen |      panel        | zen      |
 * | col |      scroll       | trailing |
 * |     |                   | col      |
 * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * |     | frozenTrailingRow |          |  frozenTrailingRowCount
 * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * @description returns which group data cell belongs in frozen mode
 */
export const getFrozenDataCellType = (
  meta: {
    colIndex: number;
    rowIndex: number;
  },
  frozenOpts: FrozenOpts,
  colLength: number,
  cellRange: {
    start: number;
    end: number;
  },
) => {
  const {
    frozenColCount,
    frozenRowCount,
    frozenTrailingColCount,
    frozenTrailingRowCount,
  } = frozenOpts;
  const { colIndex, rowIndex } = meta;
  if (isFrozenRow(rowIndex, cellRange.start, frozenRowCount)) {
    return FrozenCellType.ROW;
  }

  if (isFrozenTrailingRow(rowIndex, cellRange.end, frozenTrailingRowCount)) {
    return FrozenCellType.TRAILING_ROW;
  }

  if (isFrozenCol(colIndex, frozenColCount)) {
    return FrozenCellType.COL;
  }

  if (isFrozenTrailingCol(colIndex, frozenTrailingColCount, colLength)) {
    return FrozenCellType.TRAILING_COL;
  }
  return FrozenCellType.SCROLL;
};

/**
 * @description calculate all cells in frozen group's intersection region
 */
export const calculateFrozenCornerCells = (
  opts: FrozenOpts,
  colLength: number,
  cellRange: {
    start: number;
    end: number;
  },
) => {
  const {
    frozenColCount,
    frozenRowCount,
    frozenTrailingColCount,
    frozenTrailingRowCount,
  } = opts;

  const result: {
    [key: string]: FrozenCellIndex[];
  } = {
    [FrozenCellType.TOP]: [],
    [FrozenCellType.BOTTOM]: [],
  };

  // frozenColGroup with frozenRowGroup or frozenTrailingRowGroup. Top left and bottom left corner.
  for (let i = 0; i < frozenColCount; i++) {
    for (let j = cellRange.start; j < cellRange.start + frozenRowCount; j++) {
      result[FrozenCellType.TOP].push({
        x: i,
        y: j,
      });
    }

    if (frozenTrailingRowCount > 0) {
      for (let j = 0; j < frozenTrailingRowCount; j++) {
        const index = cellRange.end - j;
        result[FrozenCellType.BOTTOM].push({
          x: i,
          y: index,
        });
      }
    }
  }

  // frozenTrailingColGroup with frozenRowGroup or frozenTrailingRowGroup. Top right and bottom right corner.
  for (let i = 0; i < frozenTrailingColCount; i++) {
    const colIndex = colLength - 1 - i;
    for (let j = cellRange.start; j < cellRange.start + frozenRowCount; j++) {
      result[FrozenCellType.TOP].push({
        x: colIndex,
        y: j,
      });
    }

    if (frozenTrailingRowCount > 0) {
      for (let j = 0; j < frozenTrailingRowCount; j++) {
        const index = cellRange.end - j;
        result[FrozenCellType.BOTTOM].push({
          x: colIndex,
          y: index,
        });
      }
    }
  }

  return result;
};

export const isFrozenCell = (
  colIndex: number,
  rowIndex: number,
  frozenOpts: FrozenOpts,
  colLength: number,
  cellRange: {
    start: number;
    end: number;
  },
) => {
  const {
    frozenColCount,
    frozenRowCount,
    frozenTrailingColCount,
    frozenTrailingRowCount,
  } = frozenOpts;
  return (
    isFrozenCol(colIndex, frozenColCount) ||
    isFrozenTrailingCol(colIndex, frozenTrailingColCount, colLength) ||
    isFrozenRow(rowIndex, cellRange.start, frozenRowCount) ||
    isFrozenTrailingRow(rowIndex, cellRange.end, frozenTrailingRowCount)
  );
};

/**
 * @description split all cells in current panel with five child group
 */
export const splitInViewIndexesWithFrozen = (
  indexes: Indexes,
  frozenOpts: FrozenOpts,
  colLength: number,
  cellRange: {
    start: number;
    end: number;
  },
) => {
  const {
    frozenColCount,
    frozenRowCount,
    frozenTrailingColCount,
    frozenTrailingRowCount,
  } = frozenOpts;

  const centerIndexes: Indexes = [...indexes];

  // Cut off frozen cells from centerIndexes
  if (isFrozenCol(centerIndexes[0], frozenColCount)) {
    centerIndexes[0] = frozenColCount;
  }

  if (
    isFrozenTrailingCol(centerIndexes[1], frozenTrailingColCount, colLength)
  ) {
    centerIndexes[1] = colLength - frozenTrailingColCount - 1;
  }

  if (isFrozenRow(centerIndexes[2], cellRange.start, frozenRowCount)) {
    centerIndexes[2] = cellRange.start + frozenRowCount;
  }

  if (
    isFrozenTrailingRow(centerIndexes[3], cellRange.end, frozenTrailingRowCount)
  ) {
    centerIndexes[3] = cellRange.end - frozenTrailingRowCount;
  }

  // Calculate indexes for four frozen groups
  const frozenRowIndexes: Indexes = [...centerIndexes];
  frozenRowIndexes[2] = cellRange.start;
  frozenRowIndexes[3] = cellRange.start + frozenRowCount - 1;

  const frozenColIndexes: Indexes = [...centerIndexes];
  frozenColIndexes[0] = 0;
  frozenColIndexes[1] = frozenColCount - 1;

  const frozenTrailingRowIndexes: Indexes = [...centerIndexes];
  frozenTrailingRowIndexes[2] = cellRange.end + 1 - frozenTrailingRowCount;
  frozenTrailingRowIndexes[3] = cellRange.end;

  const frozenTrailingColIndexes: Indexes = [...centerIndexes];
  frozenTrailingColIndexes[0] = colLength - frozenTrailingColCount;
  frozenTrailingColIndexes[1] = colLength - 1;

  return {
    center: centerIndexes,
    frozenRow: frozenRowIndexes,
    frozenCol: frozenColIndexes,
    frozenTrailingCol: frozenTrailingColIndexes,
    frozenTrailingRow: frozenTrailingRowIndexes,
  };
};

export const getCellRange = (
  viewCellHeights: ViewCellHeights,
  pagination?: Pagination,
) => {
  const heights = viewCellHeights;
  let start = 0;
  let end = heights.getTotalLength() - 1;

  if (pagination) {
    const { current, pageSize } = pagination;

    start = Math.max((current - 1) * pageSize, 0);
    end = Math.min(current * pageSize - 1, heights.getTotalLength() - 1);
  }

  return {
    start,
    end,
  };
};
