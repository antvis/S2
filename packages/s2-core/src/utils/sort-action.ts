import { keys, has, map, toUpper, endsWith, uniq } from 'lodash';
import { SortMethod, SortParam } from '@/common/interface';
import { DataType, SortActionParams } from '@/data-set/interface';
import { EXTRA_FIELD, ID_SEPARATOR, TOTAL_VALUE } from '@/common/constant';
import { sortByItems, getListBySorted } from '@/utils/data-set-operate';
import { getDimensionsWithParentPath } from '@/utils/dataset/pivot-data-set';

export const isAscSort = (sortMethod) => toUpper(sortMethod) === 'ASC';

export const isDescSort = (sortMethod) => toUpper(sortMethod) === 'DESC';

/**
 * 执行排序
 * @param list - 待排序数组
 * @param sortMethod - 升、降序
 * @param key - 根据key数值排序，如果有key代表根据维度值排序，故按数字排，如果没有按照字典排
 */
export const sortAction = (
  list: Array<string | number | DataType>,
  sortMethod?: SortMethod,
  key?: string,
) => {
  const sort = isAscSort(sortMethod) ? 1 : -1;
  const specialValues = ['-', undefined];
  return list?.sort(
    (pre: string | number | DataType, next: string | number | DataType) => {
      let a = pre;
      let b = next;
      if (key) {
        a = pre[key];
        b = next[key];
        if (Number(a) && Number(b)) {
          return (Number(a) - Number(b)) * sort;
        }
        if (a && specialValues?.includes(a?.toString())) {
          return -sort;
        }
        if (Number(a) && specialValues?.includes(b?.toString())) {
          return sort;
        }
      }
      if (a && b) {
        // 数据健全兼容，用户数据不全时，能够展示.
        return a.toString().localeCompare(b.toString(), 'zh') * sort;
      }
      if (a) {
        return sort;
      }
      return -sort;
    },
  );
};

const mergeDataWhenASC = (
  sortedValues: string[],
  originValues: string[],
  asc: boolean,
) => {
  if (asc) {
    // 如果是升序，需要将无数据的项放到前面
    return sortByItems(originValues, uniq(sortedValues));
  }
  return [...new Set([...sortedValues, ...originValues])];
};

export const sortByFunc = (params: SortActionParams): string[] => {
  const { originValues, measureValues, sortParam } = params;
  const { sortFunc } = sortParam;
  return (
    (sortFunc({ data: measureValues, ...sortParam }) as string[]) ||
    originValues
  );
};

export const sortByCustom = (params: SortActionParams): string[] => {
  const { sortByValues, originValues } = params;

  // 从 originValues 中过滤出所有包含 sortByValue 的 id
  const idWithPre = originValues.filter((originItem) =>
    sortByValues.find((value) => endsWith(originItem, value)),
  );
  // 将 id 拆分为父节点和目标节点
  const idListWithPre = idWithPre.map((idStr) => {
    const ids = idStr.split(ID_SEPARATOR);
    if (ids.length > 1) {
      const parentId = ids.slice(0, ids.length - 1).join(ID_SEPARATOR);
      return [parentId, ids[ids.length - 1]];
    }
    return ids;
  });
  // 获取父节点顺序
  const parentOrder = Array.from(new Set(idListWithPre.map((id) => id[0])));
  // 排序
  idListWithPre.sort((a: string[], b: string[]) => {
    const aParent = a.slice(0, a.length - 1);
    const bParent = b.slice(0, b.length - 1);
    // 父节点不同时，按 parentOrder 排序
    if (aParent.join() !== bParent.join()) {
      const aParentIndex = parentOrder.indexOf(aParent[0]);
      const bParentIndex = parentOrder.indexOf(bParent[0]);
      return aParentIndex - bParentIndex;
    }
    // 父节点相同时，按 sortByValues 排序
    const aIndex = sortByValues.indexOf(a[a.length - 1]);
    const bIndex = sortByValues.indexOf(b[b.length - 1]);
    return aIndex - bIndex;
  });
  // 拼接 id
  const sortedIdWithPre = idListWithPre.map((idArr) =>
    idArr.join(ID_SEPARATOR),
  );

  return getListBySorted(originValues, sortedIdWithPre);
};

export const sortByMethod = (params: SortActionParams): string[] => {
  const { sortParam, measureValues, originValues, dataSet } = params;
  const { sortByMeasure, query, sortFieldId, sortMethod } = sortParam;
  const { rows, columns } = dataSet.fields;
  const isInRows = rows.includes(sortFieldId);
  let result = originValues;

  if (sortByMeasure) {
    const dimensions = sortAction(
      measureValues,
      sortMethod,
      sortByMeasure === TOTAL_VALUE ? query[EXTRA_FIELD] : sortByMeasure,
    ) as Record<string, DataType>[];

    result = getDimensionsWithParentPath(
      sortFieldId,
      isInRows ? rows : columns,
      dimensions,
    );
  } else {
    result = map(sortAction(measureValues, sortMethod)) as string[];
  }

  return mergeDataWhenASC(result, originValues, isAscSort(sortMethod));
};

export const processSort = (params: SortActionParams): string[] => {
  const { sortParam, originValues, measureValues, dataSet } = params;
  const { sortFunc, sortMethod, sortBy } = sortParam;

  let result = originValues;
  const sortActionParams = {
    originValues,
    measureValues,
    sortParam,
    dataSet,
  };
  if (sortFunc) {
    result = sortByFunc(sortActionParams);
  } else if (sortBy) {
    // 自定义列表
    result = sortByCustom({ sortByValues: sortBy, originValues });
  } else if (isAscSort(sortMethod) || isDescSort(sortMethod)) {
    // 如果是升序，需要将无数据的项放到前面
    result = sortByMethod(sortActionParams);
  }
  return result;
};

export const handleSortAction = (params: SortActionParams): string[] => {
  const { dataSet, sortParam, originValues, isSortByMeasure } = params;
  const { fields } = dataSet;
  const { sortByMeasure, query, sortFieldId } = sortParam;
  let measureValues;
  if (isSortByMeasure) {
    // 根据指标排序，需要首先找到指标的对应的值
    if (sortByMeasure === TOTAL_VALUE) {
      // 按小计，总计排序
      const isRow =
        fields?.columns?.includes(sortFieldId) &&
        keys(query)?.length === 1 &&
        has(query, EXTRA_FIELD);
      measureValues = dataSet.getMultiData(query, true, isRow);
    } else {
      measureValues = dataSet.getMultiData(query);
    }
  } else {
    // 其他都是维度本身的排序方式
    measureValues = originValues;
  }
  return processSort({
    sortParam,
    originValues,
    measureValues,
    dataSet,
  });
};

export const getSortTypeIcon = (sortParam: SortParam, isSortCell?: boolean) => {
  if (sortParam?.sortMethod) {
    if (isAscSort(sortParam?.sortMethod)) {
      return 'groupAsc';
    }
    if (isDescSort(sortParam?.sortMethod)) {
      return 'groupDesc';
    }
  }
  if (isSortCell) {
    return 'SortDown';
  }
};
