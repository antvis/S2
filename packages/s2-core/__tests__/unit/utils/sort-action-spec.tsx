import { sortAction, sortByCustom } from '@/utils/sort-action';

describe('Sort Action Test', () => {
  describe('Sort Action', () => {
    test('sort action with number arr', () => {
      const data = [1, 3, 2];
      expect(sortAction(data, 'ASC')).toEqual([1, 2, 3]);
      expect(sortAction(data, 'DESC')).toEqual([3, 2, 1]);
    });

    test('sort action with number-string and number arr', () => {
      const data1 = ['11', '3', 2];
      expect(sortAction(data1, 'ASC')).toEqual(['11', 2, '3']);
      expect(sortAction(data1, 'DESC')).toEqual(['3', 2, '11']);

      const data2 = ['11', '3', '2'];
      expect(sortAction(data2, 'ASC')).toEqual(['11', '2', '3']);
      expect(sortAction(data2, 'DESC')).toEqual(['3', '2', '11']);
    });

    test('sort action with string arr', () => {
      const data = ['a', 'c', 'b'];
      expect(sortAction(data, 'ASC')).toEqual(['a', 'b', 'c']);
      expect(sortAction(data, 'DESC')).toEqual(['c', 'b', 'a']);

      const data1 = ['啊', '哦', '嗯'];
      expect(sortAction(data1, 'ASC')).toEqual(['啊', '嗯', '哦']);
      expect(sortAction(data1, 'DESC')).toEqual(['哦', '嗯', '啊']);

      const data2 = ['啊', '11', '2'];
      expect(sortAction(data2, 'ASC')).toEqual(['11', '2', '啊']);
      expect(sortAction(data2, 'DESC')).toEqual(['啊', '2', '11']);
    });

    test('sort action with object arr', () => {
      const data1 = [{ a: 1 }, { a: 3 }, { a: 2 }];
      expect(sortAction(data1, 'ASC', 'a')).toEqual([
        { a: 1 },
        { a: 2 },
        { a: 3 },
      ]);
      expect(sortAction(data1, 'DESC', 'a')).toEqual([
        { a: 3 },
        { a: 2 },
        { a: 1 },
      ]);

      const data2 = [{ a: '11' }, { a: '3' }, { a: 2 }];
      expect(sortAction(data2, 'ASC', 'a')).toEqual([
        { a: 2 },
        { a: '3' },
        { a: '11' },
      ]);
      expect(sortAction(data2, 'DESC', 'a')).toEqual([
        { a: '11' },
        { a: '3' },
        { a: 2 },
      ]);

      const data3 = [{ a: '-' }, { a: '3' }, { a: 2 }];
      expect(sortAction(data3, 'ASC', 'a')).toEqual([
        { a: '-' },
        { a: 2 },
        { a: '3' },
      ]);
      expect(sortAction(data3, 'DESC', 'a')).toEqual([
        { a: '3' },
        { a: 2 },
        { a: '-' },
      ]);

      expect(
        sortAction(
          [{ a: '-' }, { a: '3' }, { a: 2 }, { a: undefined }],
          'ASC',
          'a',
        ),
      ).toEqual([{ a: undefined }, { a: '-' }, { a: 2 }, { a: '3' }]);
      expect(
        sortAction(
          [{ a: '-' }, { a: '3' }, { a: 2 }, { a: undefined }],
          'DESC',
          'a',
        ),
      ).toEqual([{ a: '3' }, { a: 2 }, { a: '-' }, { a: undefined }]);
      expect(sortAction([{ a: '' }, { a: '3' }, { a: 2 }], 'ASC', 'a')).toEqual(
        [{ a: '' }, { a: 2 }, { a: '3' }],
      );
    });
  });
});

describe('Sort By Custom Test', () => {
  describe('Sort By Custom', () => {
    test('sort by custom with equal sub node', () => {
      const params = {
        originValues: [
          'Monday[&]noon',
          'Monday[&]afternoon',
          'Monday[&]morning',
          'Tuesday[&]afternoon',
          'Tuesday[&]noon',
          'Tuesday[&]morning',
        ],
        sortByValues: ['morning', 'noon', 'afternoon'],
      };
      expect(sortByCustom(params)).toEqual([
        'Monday[&]morning',
        'Monday[&]noon',
        'Monday[&]afternoon',
        'Tuesday[&]morning',
        'Tuesday[&]noon',
        'Tuesday[&]afternoon',
      ]);
    });
    test('sort by custom with repeated sub node', () => {
      const params = {
        originValues: [
          'Monday[&]noon',
          'Monday[&]afternoon',
          'Tuesday[&]afternoon',
          'Tuesday[&]noon',
          'Tuesday[&]morning',
          'Wednesday[&]afternoon',
          'Wednesday[&]morning',
        ],
        sortByValues: ['morning', 'noon', 'afternoon'],
      };
      expect(sortByCustom(params)).toEqual([
        'Monday[&]noon',
        'Monday[&]afternoon',
        'Tuesday[&]morning',
        'Tuesday[&]noon',
        'Tuesday[&]afternoon',
        'Wednesday[&]morning',
        'Wednesday[&]afternoon',
      ]);
    });
    test('sort by custom with unordered node', () => {
      const params = {
        originValues: [
          'Monday[&]afternoon',
          'Tuesday[&]afternoon',
          'Wednesday[&]afternoon',
          'Monday[&]noon',
          'Tuesday[&]noon',
          'Wednesday[&]morning',
          'Tuesday[&]morning',
        ],
        sortByValues: ['morning', 'noon', 'afternoon'],
      };
      expect(sortByCustom(params)).toEqual([
        'Monday[&]noon',
        'Monday[&]afternoon',
        'Tuesday[&]morning',
        'Tuesday[&]noon',
        'Tuesday[&]afternoon',
        'Wednesday[&]morning',
        'Wednesday[&]afternoon',
      ]);
    });
  });
});
