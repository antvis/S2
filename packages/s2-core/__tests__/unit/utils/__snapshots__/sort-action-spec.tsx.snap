// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GetSortByMeasureValues Total Fallback Tests should sort by col total whit group 1`] = `
Array [
  Object {
    "$$extra$$": "price",
    "$$value$$": "666",
    "city": "杭州",
    "price": "666",
    "type": "笔",
  },
  Object {
    "$$extra$$": "price",
    "$$value$$": "999",
    "city": "杭州",
    "price": "999",
    "type": "纸张",
  },
]
`;
